cf/data/data.py

```diff
@@ -830,78 +830,75 @@ def __iter__(self):
+
+        If the shape of the data is unknown then it is calculated
+        immediately by executing all delayed operations.
+
+        **Examples**
 
         >>> d = cf.Data([1, 2, 3], 'metres')
         >>> for e in d:
-        ...    print(repr(e))
+        ...     print(repr(e))
         ...
-        1
-        2
-        3
+        <CF Data(1): [1] metres>
+        <CF Data(1): [2] metres>
+        <CF Data(1): [3] metres>
 
-        >>> d = cf.Data([[1, 2], [4, 5]], 'metres')
+        >>> d = cf.Data([[1, 2], [3, 4]], 'metres')
         >>> for e in d:
-        ...    print(repr(e))
+        ...     print(repr(e))
         ...
         <CF Data: [1, 2] metres>
-        <CF Data: [4, 5] metres>
+        <CF Data: [3, 4] metres>
 
-        >>> d = cf.Data(34, 'metres')
+        >>> d = cf.Data(99, 'metres')
         >>> for e in d:
         ...     print(repr(e))
-        ..
+        ...
+        Traceback (most recent call last):
+            ...
         TypeError: iteration over a 0-d Data
 
         """
-        ndim = self._ndim
-
-        if not ndim:
-            raise TypeError(
-                "Iteration over 0-d {}".format(self.__class__.__name__)
-            )
+        try:
+            n = len(self)
+        except TypeError:
+            raise TypeError(f"iteration over a 0-d {self.__class__.__name__}")
 
-        elif ndim == 1:
-            if self.fits_in_memory(self.dtype.itemsize):
-                i = iter(self.array)
-                while 1:
-                    try:
-                        yield next(i)
-                    except StopIteration:
-                        return
-            else:
-                for n in range(self._size):
-                    yield self[n].array[0]
-
-        else:
-            # ndim > 1
-            for n in range(self._shape[0]):
-                out = self[n, ...]
-                out.squeeze(0, inplace=True)
-                yield out
+        for i in range(n):
+            yield self[i]
 
     def __len__(self):
-        """The built-in function `len`
+        """Called to implement the built-in function `len`.
 
         x.__len__() <==> len(x)
 
-        **Examples:**
+        **Performance**
+
+        If the shape of the data is unknown then it is calculated
+        immediately by executing all delayed operations.
 
-        >>> len(Data([1, 2, 3]))
+        **Examples**
+
+        >>> len(cf.Data([1, 2, 3]))
         3
-        >>> len(Data([[1, 2, 3]]))
+        >>> len(cf.Data([[1, 2, 3]]))
         1
-        >>> len(Data([[1, 2, 3], [4, 5, 6]]))
+        >>> len(cf.Data([[1, 2, 3], [4, 5, 6]]))
         2
-        >>> len(Data(1))
-        TypeError: len() of scalar Data
+        >>> len(cf.Data(1))
+        Traceback (most recent call last):
+            ...
+        TypeError: len() of unsized object
 
         """
-        shape = self._shape
-        if shape:
-            return shape[0]
+        dx = self._get_dask()
+        if math.isnan(dx.size):
```

Comment:
`isnan` seems to be the only method we use from `math` across the Data module and to avoid this for full use of `numpy`/`dask` over `math`, perhaps with some performance benefit, can we not instead use (here and the other methods  which use it, e.g. `size`, `shape`, `nbytes`, etc.):

```suggestion
        if da.isnan(dx.size):
```
(else `np.isnan`)?

Reply:
It turns out that dask itself seems very confused and uses math.isnan and np.isnan seemingly randomly for scalars (you need np.isnan for array checks) - I also just chose one at random:

```shell
$ grep isnan dask/array/slicing.py 
        if np.isnan(dim) and ind != slice(None, None, None):
    if math.isnan(other_numel) or other_numel == 0:
        if ind < 0 and not math.isnan(shape):
    if isinstance(ind, (np.ndarray, list)) and not math.isnan(shape):
        if math.isnan(dim):
    if np.isnan(dimension):
    if np.isnan(x.chunks[axis]).any():
        if not np.isnan(x.shape).any() and not np.isnan(index[0].shape).any():
                and not math.isnan(index_size)
            if np.isnan(index.size):
            if math.isnan(b) or b <= index.size:
        elif math.isnan(a):
$ 
```
A quick performance test shows that for scalars `math.isnan` is ~8x faster than np.isnan, and ~15x faster than da.isnan (which itself is just  a wrapper to np.isnan). So (by luck rather than judgement!) I think it probably is worth going with math.isnan. Is that OK?